shaping/output.go

```diff
@@ -183,40 +200,131 @@ func (o *Output) advanceSpaceAware() fixed.Int26_6 {
 	var (
 		advance fixed.Int26_6
-		tallest fixed.Int26_6
-		lowest  fixed.Int26_6
+		ascent  fixed.Int26_6
+		descent fixed.Int26_6
 	)
 
 	if o.Direction.IsVertical() {
 		for i := range o.Glyphs {
 			g := &o.Glyphs[i]
 			advance += g.YAdvance
 			depth := g.XOffset + g.XBearing // start of the glyph
-			if depth < lowest {
-				lowest = depth
+			if depth < descent {
+				descent = depth
 			}
-			height := g.XOffset + g.Width // end of the glyph
-			if height > tallest {
-				tallest = height
+			height := depth + g.Width // end of the glyph
+			if height > ascent {
+				ascent = height
 			}
 		}
 	} else { // horizontal
 		for i := range o.Glyphs {
 			g := &o.Glyphs[i]
 			advance += g.XAdvance
 			height := g.YBearing + g.YOffset
-			if height > tallest {
-				tallest = height
+			if height > ascent {
+				ascent = height
 			}
 			depth := height + g.Height
-			if depth < lowest {
-				lowest = depth
+			if depth < descent {
+				descent = depth
 			}
 		}
 	}
 	o.Advance = advance
 	o.GlyphBounds = Bounds{
-		Ascent:  tallest,
-		Descent: lowest,
+		Ascent:  ascent,
+		Descent: descent,
+	}
+}
+
+// Assuming [Glyphs] comes from an horizontal shaping,
+// applies a 90°, clockwise rotation to the whole slice of glyphs,
+// to create 'sideways' vertical text.
+//
+// The [Direction] field is updated by switching the axis to vertical
+// and the orientation to "sideways".
+//
+// [RecalculateAll] should be called afterwards to update [Avance] and [GlyphBounds].
+func (out *Output) sideways() {
+	for i, g := range out.Glyphs {
+		// switch height and width
+		out.Glyphs[i].Width = -g.Height // height is negative
+		out.Glyphs[i].Height = -g.Width
+		// compute the bearings
+		out.Glyphs[i].XBearing = g.YBearing + g.Height
+		out.Glyphs[i].YBearing = g.Width
+		// switch advance direction
+		out.Glyphs[i].XAdvance = 0
+		out.Glyphs[i].YAdvance = -g.XAdvance // YAdvance is negative
+		// apply a rotation around the dot, and position the glyph
+		// below the dot
+		out.Glyphs[i].XOffset = g.YOffset
+		out.Glyphs[i].YOffset = -(g.XOffset + g.XBearing + g.Width)
+	}
+
+	// adjust direction
+	out.Direction.SetSideways(true)
+}
+
+// properly update [GlyphBounds]
+func (out *Output) moveCrossAxis(d fixed.Int26_6) {
+	if out.Direction.IsVertical() {
+		for i := range out.Glyphs {
+			out.Glyphs[i].XOffset += d
+		}
+	} else {
+		for i := range out.Glyphs {
+			out.Glyphs[i].YOffset += d
+		}
+	}
+	out.GlyphBounds.Ascent += d
+	out.GlyphBounds.Descent += d
+}
+
+// AdjustBaseline aligns runs with different baselines.
+//
+// For vertical text, it centralizes 'sideways' runs, so
+// that text with mixed 'upright' and
+// 'sideways' orientation is better aligned.
+//
+// This is currently a no-op for horizontal text.
+//
+// Note that this method only update cross-axis metrics,
+// so that the advance is preserved. As such, it is valid
+// to call this method after line wrapping, for instance.
+func (l Line) AdjustBaseline() {
```

Comment:
Reading the doc implies this should be called `AlignBaselines` instead.
Just a thought, not a change request.

Reply:
I've added an "s" to imply each run may be shifted.